tilings/algorithms/obstruction_inferral.py

```diff
@@ -0,0 +1,149 @@
+        newobs = []
+        merged_tiling = self._tiling.merge()
+        for ob in sorted(self.potential_new_obs(), key=len):
+            if self.can_add_obstruction(ob, merged_tiling):
+                newobs.append(ob)
+                merged_tiling = merged_tiling.__class__(
+                    obstructions=chain(merged_tiling.obstructions, (ob,)),
+                    requirements=merged_tiling.requirements,
+                    remove_empty=False)
+        self._new_obs = newobs
+        return self._new_obs
+
+    @staticmethod
+    def can_add_obstruction(obstruction, tiling):
+        """Return true if obstruction can be added to tiling."""
+        return (tiling.add_requirement(obstruction.patt, obstruction.pos)
+                .merge().is_empty())
+
+    def obstruction_inferral(self):
+        """
+        Return the tiling with the new obstructions.
+        """
+        obs = chain(self._tiling.obstructions, self.new_obs())
+        reqs = self._tiling.requirements
+        return self._tiling.__class__(obs, reqs)
+
+    def formal_step(self):
+        """ Return a string describing the operation performed. """
+        return "Added the obstructions {}.".format(self.new_obs())
+
+    def rule(self):
+        """
+        Return a comb_spec_searcher Rule for the new tiling.
+
+        If no new obstruction is added, returns None.
+        """
+        if self.new_obs():
+            return InferralRule(self.formal_step(),
+                                self.obstruction_inferral())
+
+
+class SubobstructionInferral(ObstructionInferral):
+    """
+    Algorithm to compute the tiling created by adding all
+    subobstructions which can be added.
+    """
+
+    def potential_new_obs(self):
+        """
+        Return the set of all subobstructions of the tiling.
+        """
+        subobs = set()
+        for ob in self._tiling.obstructions:
+            subobs.update(ob.all_subperms())
+        subobs.remove(Obstruction(Perm(), []))
+        return subobs
+
+
+class AllObstructionInferral(ObstructionInferral):
+    """
+    Algorithm to compute the tiling created by adding all
+    obstruction of given length which can be added.
+    """
+    def __init__(self, tiling, obstrucion_length):
+        super().__init__(tiling)
+        self._obs_len = obstrucion_length
+
+    @property
+    def obstrucion_length(self):
+        return self._obs_len
+
+    def avoids_obstructions(self, gp):
+        """
+        Return True if the gridded perm `gp` does not contains any obstruction
+        of the tiling.
+        """
+        return all(ob not in gp for ob in self._tiling.obstructions)
+
+    def not_required(self, gp):
+        """
+        Returns True if the gridded perm `gp` is not required by any
+        requirement list of the tiling.
+        """
+        return all(any(gp not in req for req in req_list)
+                   for req_list in self._tiling.requirements)
+
+    def potential_new_obs(self):
```

Comment:
I would suggest doing 
```
yield from Tiling(obstructions).gridded_perms_of_length(length)
```
It should achieve the same thing with far less explosion.

Reply:
I think that this suggestion doesn't work. Take the following code
```python
t = Tiling(obstructions=[
    Obstruction(Perm((0, 1, 2)), ((0, 0),)*3),
    Obstruction(Perm((0, 1, 2)), ((1, 0),)*3),
    Obstruction(Perm((0, 1)), ((0, 0), (1, 0))),
    Obstruction(Perm((1, 0)), ((0, 0), (1, 0))),
], requirements=[[
    Requirement(Perm((0,)), ((1, 0),)),
]])
print(t)
print('Contained gridded permutations of length 1')
for p in t.gridded_perms_of_length(1):
    print(p)
```
The output is
```
+-+-+
|1|2|
+-+-+
1: Av(012)
2: Av+(012)
Crossing obstructions:
01: (0, 0), (1, 0)
10: (0, 0), (1, 0)
Requirement 0:
0: (1, 0)
Contained gridded permutations of length 1
0: (1, 0)
```
With your suggestion we would never try to add the obstruction `0: (0, 0)`

Reply:
We would, note, I am suggesting initialising without the requirements.

Reply:
A thought: 

Let T = (O, R) be a tiling, and T' = (O, emptyset).

Is Grid(T') - Grid(T) precisely those we want to add?

Reply:
This is obviously wrong... (e.g every one by one positive tiling becomes empty...). 

However, all obstructions we can add do satisfy o in Grid(T') and o not in Grid(T).

Reply:
I pushed a more clever filtering that includes what we've discussed